[FSDP()] Fix `fully_shard` fwd hook registration

awgu opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#90201 [FSDP()] Fix `fully_shard` fwd hook registration**

I need to rebase later after Shen's PRs land.

The idea is to only register the pre/post-forward hook on the _root modules_ among the modules that consume a `FlatParameter`. (Yes, the term _root module_ is heavily overloaded. We may want to clarify that at some point. Here, _root_ is being used in the graph sense, meaning parent-less, and the scope is only among the modules consuming a `FlatParameter`.)

This avoids unnecessary pre/post-forward hooks running, which would lead to errors because the unshard is not truly idempotent.

Commit:
[FSDP()] Fix `fully_shard` fwd hook registration

[ghstack-poisoned]

Commit:
Update on "[FSDP()] Fix `fully_shard` fwd hook registration"


I need to rebase later after Shen's PRs land.

[ghstack-poisoned]

awgu commented on torch/distributed/fsdp/_init_utils.py:
```diff
@@ -175,10 +184,6 @@ def _init_runtime_state(
     state._pre_forward_handles = _pre_forward_handles
     _post_forward_handles: List[RemovableHandle] = []
     state._post_forward_handles = _post_forward_handles
-    _module_to_handles: Dict[
```
Remove redundant definition that slipped through

mrshenli commented on torch/distributed/fsdp/_common_utils.py:
```diff
@@ -200,3 +200,27 @@ def _assert_in_training_states(
+def _get_root_modules(modules: Set[nn.Module]) -> Set[nn.Module]:
+    """
+    Returns:
+        Set[nn.Module: The subset of ``modules`` that are root modules (i.e.
```
missing "]"?

awgu replied:
Fixed

mrshenli reviewed (approved):
LGTM! Thanks for promptly fixing this!

Commit:
Update on "[FSDP()] Fix `fully_shard` fwd hook registration"


I need to rebase later after Shen's PRs land.

The idea is to only register the pre/post-forward hook on the _root modules_ among the modules that consume a `FlatParameter`. (Yes, the term _root module_ is heavily overloaded. We may want to clarify that at some point. Here, _root_ is being used in the graph sense, meaning parent-less, and the scope is only among the modules consuming a `FlatParameter`.)

This avoids unnecessary pre/post-forward hooks running, which would lead to errors because the unshard is not truly idempotent.

[ghstack-poisoned]

rohan-varma commented on torch/distributed/fsdp/flat_param.py:
```diff
@@ -265,6 +268,7 @@ def _init_metadata(
         self._modules = set(pi.module for pi in self._param_infos).union(
             set(spi.module for spi in self._shared_param_infos)
         )
+        self._root_modules = _get_root_modules(self._modules)
```
nit: might be useful to add a comment describing why this property is needed (same goes for other properties as well).

awgu replied:
Added explanation in the `FlatParameter` docstring

rohan-varma reviewed (approved):
LGTM overall, but could we add a unittest that would've captured this issue?

awgu commented:
> LGTM overall, but could we add a unittest that would've captured this issue?

Fortunately, Shen added the failing unit test in his PR (and the PR below this one patched that), so we have a unit test for this :)

Commit:
Update on "[FSDP()] Fix `fully_shard` fwd hook registration"


I need to rebase later after Shen's PRs land.

The idea is to only register the pre/post-forward hook on the _root modules_ among the modules that consume a `FlatParameter`. (Yes, the term _root module_ is heavily overloaded. We may want to clarify that at some point. Here, _root_ is being used in the graph sense, meaning parent-less, and the scope is only among the modules consuming a `FlatParameter`.)

This avoids unnecessary pre/post-forward hooks running, which would lead to errors because the unshard is not truly idempotent.

[ghstack-poisoned]

Commit:
Update on "[FSDP()] Fix `fully_shard` fwd hook registration"


I need to rebase later after Shen's PRs land.

The idea is to only register the pre/post-forward hook on the _root modules_ among the modules that consume a `FlatParameter`. (Yes, the term _root module_ is heavily overloaded. We may want to clarify that at some point. Here, _root_ is being used in the graph sense, meaning parent-less, and the scope is only among the modules consuming a `FlatParameter`.)

This avoids unnecessary pre/post-forward hooks running, which would lead to errors because the unshard is not truly idempotent.

[ghstack-poisoned]
